identity/src/main/java/com/android/identity/GattServer.java

```diff
@@ -209,25 +268,103 @@ public void onCharacteristicReadRequest(BluetoothDevice device, int requestId, i
+            if (mLog.isTransportEnabled()) {
+                mLog.transport("PSM Value generated: " + psm + " byte: " + Util.toHex(psmValue));
+            }
+            mGattServer.sendResponse(device,
+                    requestId,
+                    BluetoothGatt.GATT_SUCCESS,
+                    0,
+                    psmValue);
+            acceptConnection();
         } else {
             reportError(new Error("Read on unexpected characteristic with UUID "
                     + characteristic.getUuid()));
         }
     }
 
+    private void acceptConnection() {
+        try {
+            mSocket = mServerSocket.accept();
+            if (mSocket.isConnected()) {
+                // Stop accepting new connections
+                mServerSocket.close();
+                reportPeerConnected();
+                listenSocket();
+            }
+        } catch (IOException e) {
+            Log.e(TAG, "Error getting connection from socket server for L2CAP " + e.getMessage(), e);
+            reportError(new Error("Error getting connection from socket server for L2CAP"));
+        }
+    }
+
+    private void listenSocket() {
+        if (mLog.isTransportEnabled()) {
+            mLog.transport("Start reading socket input");
+        }
+        // We can check better/right buffer size to use with L2CAP
+        byte[] mmBuffer = new byte[32027];
+        int numBytes; // bytes returned from read()
+        // Keep listening to the InputStream until an exception occurs.
+        while (mSocket.isConnected()) {
+            try {
+                // Read from the InputStream.
+                numBytes = mSocket.getInputStream().read(mmBuffer, 0, mmBuffer.length);
+                if (numBytes == -1) {
+                    if (mLog.isTransportEnabled()) {
+                        mLog.transport("Stop listening from socket");
+                    }
+                    reportPeerDisconnected();
+                    break;
+                }
+                if (mLog.isTransportEnabled()) {
+                    mLog.transport("Chunk received by socket: (" + numBytes + ") " + Util.toHex(mmBuffer));
+                }
+                // Report message received.
+                mIncomingMessage.write(mmBuffer, 0, numBytes);
+                if (numBytes < mmBuffer.length) {
+                    // Last chunk received.
+                    byte[] entireMessage = mIncomingMessage.toByteArray();
+                    if (mLog.isTransportEnabled()) {
+                        mLog.transport("Entire message received by socket: (" + entireMessage.length + ") " + Util.toHex(entireMessage));
+                    }
+                    mIncomingMessage.reset();
+                    reportMessageReceived(entireMessage);
+                }
+            } catch (IOException e) {
+                reportError(new Error("Error on listening input stream from socket L2CAP"));
+                break;
+            }
+        }
```

Comment:
I'm surprised this even works. I mean, this code all runs in the thread that was used when the GATT client reads the L2CAP characteristic... through `acceptConnection()` which calls `listenSocket()`. So we get here and we seem to take over the thread for listening to data. I think it works because `onCharacteristicReadRequest()` is called by a BLE worker thread which we then monopolize for reading...

I think if we change this so you have a dedicated thread for listening for connections and then reusing same thread for reading then it will work. See `DataTransportTcp.java` methods `listen()` and `processMessagesFromSocket()` for an example of how that works...

Reply:
Thread created for L2CAP socket